File: voice-game-framework/shared/schema.py
"""Canonical protocol — Python side. Mirrors schema.ts and protocol.md.

Imported by the agent (game_context.py, game_tools.py, bot_npc.py). If you
change a field here, change schema.ts and protocol.md too.

Pydantic v2 models. Inbound game messages are parsed defensively (extra fields
ignored) so a slightly-ahead client never crashes the agent.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

SCHEMA_VERSION = 1

NPCRole = Literal["dealer", "bodyguard", "bystander"]
NPCDisposition = Literal["hostile", "neutral", "friendly", "nervous"]
NPCAnimationState = Literal["idle", "alert", "aggressive", "relaxed", "fleeing"]
NPCAction = Literal["step_forward", "draw_weapon", "holster_weapon", "flee", "relax"]


class _Lenient(BaseModel):
    # Ignore unknown fields so a newer client doesn't break the agent.
    model_config = ConfigDict(extra="ignore")


class Vec3(_Lenient):
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0


class ActionNote(_Lenient):
    type: str
    note: str | None = None


class PlayerState(_Lenient):
    position: Vec3 = Field(default_factory=Vec3)
    looking_at: str | None = None
    distance_to_dealer: float = 999.0
    is_speaking: bool = False
    movement_speed: float = 0.0
    approached_calmly: bool = True
    last_action: ActionNote | None = None


class NPCSnapshot(_Lenient):
    id: str
    name: str = ""
    role: NPCRole = "dealer"
    position: Vec3 = Field(default_factory=Vec3)
    distance_to_player: float = 999.0
    disposition: NPCDisposition = "neutral"
    is_armed: bool = False
    animation_state: NPCAnimationState = "idle"


class SceneSnapshot(_Lenient):
    id: str = "scene"
    tension_level: float = 0.0
    ambient_events: list[str] = Field(default_factory=list)


# ---- Game → Agent -------------------------------------------------------


class GameStateMessage(_Lenient):
    type: Literal["game_state"] = "game_state"
    schema_version: int = SCHEMA_VERSION
    session_id: str = ""
    conversation_id: str = ""
    timestamp_ms: int = 0
    scene: SceneSnapshot = Field(default_factory=SceneSnapshot)
    player: PlayerState = Field(default_factory=PlayerState)
    npcs: list[NPCSnapshot] = Field(default_factory=list)


# ---- Agent → Game -------------------------------------------------------


class NPCActionMessage(_Lenient):
    type: Literal["npc_action"] = "npc_action"
    npc_id: str
    action: NPCAction
    params: dict | None = None
    utterance_id: str | None = None
    reason: str | None = None


class TensionUpdateMessage(_Lenient):
    type: Literal["tension_update"] = "tension_update"
    tension_level: float


class SubtitleMessage(_Lenient):
    type: Literal["subtitle"] = "subtitle"
    npc_id: str
    text: str
    utterance_id: str | None = None

File: voice-game-framework/shared/schema.ts
// Canonical protocol — TypeScript side. Mirrors schema.py and protocol.md.
// Imported by the game client (client/pipecatBridge.ts + scene-negotiation).
// If you change a field here, change schema.py and protocol.md too.

export const SCHEMA_VERSION = 1;

export interface Vec3 {
  x: number;
  y: number;
  z: number;
}

export type NPCRole = "dealer" | "bodyguard" | "bystander";
export type NPCDisposition = "hostile" | "neutral" | "friendly" | "nervous";
export type NPCAnimationState =
  | "idle"
  | "alert"
  | "aggressive"
  | "relaxed"
  | "fleeing";
export type NPCAction =
  | "step_forward"
  | "draw_weapon"
  | "holster_weapon"
  | "flee"
  | "relax";

export interface ActionNote {
  type: string;
  note?: string;
}

export interface PlayerState {
  position: Vec3;
  looking_at: string | null;
  distance_to_dealer: number;
  is_speaking: boolean;
  movement_speed: number;
  approached_calmly: boolean;
  last_action?: ActionNote;
}

export interface NPCSnapshot {
  id: string;
  name: string;
  role: NPCRole;
  position: Vec3;
  distance_to_player: number;
  disposition: NPCDisposition;
  is_armed: boolean;
  animation_state: NPCAnimationState;
}

export interface SceneSnapshot {
  id: string;
  tension_level: number; // 0..1
  ambient_events: string[];
}

// ---- Game → Agent -------------------------------------------------------

export interface GameStateMessage {
  type: "game_state";
  schema_version: number;
  session_id: string;
  conversation_id: string;
  timestamp_ms: number;
  scene: SceneSnapshot;
  player: PlayerState;
  npcs: NPCSnapshot[];
}

export interface ConversationStartMessage {
  type: "conversation_start";
  npc_id: string;
  conversation_id: string;
}

export interface ConversationEndMessage {
  type: "conversation_end";
  npc_id: string;
  conversation_id: string;
}

export type GameToAgentMessage =
  | GameStateMessage
  | ConversationStartMessage
  | ConversationEndMessage;

// ---- Agent → Game -------------------------------------------------------

export interface NPCActionMessage {
  type: "npc_action";
  npc_id: string;
  action: NPCAction;
  params?: Record<string, number | string | boolean>;
  utterance_id?: string;
  reason?: string;
}

export interface TensionUpdateMessage {
  type: "tension_update";
  tension_level: number;
}

export interface SubtitleMessage {
  type: "subtitle";
  npc_id: string;
  utterance_id?: string;
  text: string;
}

export type AgentToGameMessage =
  | NPCActionMessage
  | TensionUpdateMessage
  | SubtitleMessage;
